File: Leetcode/java/First Missing Positive.java
public class Solution {
    public int firstMissingPositive(int[] num) {
        int n = num.length;
        
        for (int i = 0; i < n; i++) {
            while (num[i] != i + 1 && num[i] > 0 && num[i] <= n) {
                if (num[num[i] - 1] == num[i]) {
                    break;
                }
                int tmp = num[num[i] - 1];
                num[num[i] - 1] = num[i];
                num[i] = tmp;
            }
        }
        int miss = n + 1;
        for (int i = 0; i < n; i++) {
            if (num[i] != i + 1) {
                miss = i + 1;
                break;
            }
        }
        return miss;
    }
}

File: LintCode/cpp/First Missing Positive.cc
class Solution {
public:
    /**    
     * @param A: a vector of integers
     * @return: an integer
     */
    int firstMissingPositive(vector<int> A) {
        int n = A.size();
        radix_sort(A, n);
        for (int i = 0; i < n; i++) {
            if (A[i] != i + 1) {
                return i + 1;
            }
        }
        return n + 1;
    }
private:
    void radix_sort(vector<int>& A, int n) {
        for (int i = 0; i < n; i++) {
            while (true) {
                if (A[i] <= 0 || A[i] > n) {
                    break;
                }
                if (A[A[i] - 1] == A[i]) {
                    break;
                }
                swap(A[A[i] - 1], A[i]);
            }
        }
    }
};
